icicle_v3/backend/cpu/include/cpu_ntt.h

```diff
@@ -0,0 +1,331 @@
+          for (int j = 0; j < half_len; ++j) {
+            int tw_idx = (dir == NTTDir::kForward) ? j * step : domain_max_size - j * step;
+            E u = current_elements[i + j];
+            E v = current_elements[i + j + half_len] * twiddles[tw_idx];
+            current_elements[i + j] = u + v;
+            current_elements[i + j + half_len] = u - v;
+          }
+        }
+      }
+    }
+  }
+
+  template <typename S = scalar_t, typename E = scalar_t>
+  void dif_ntt(E* elements, int size, int batch_size, const S* twiddles, NTTDir dir, int domain_max_size)
+  {
+    for (int batch = 0; batch < batch_size; ++batch) {
+      E* current_elements = elements + batch * size;
+      for (int len = size; len >= 2; len >>= 1) {
+        int half_len = len / 2;
+        int step = (size / len) * (domain_max_size / size);
+        for (int i = 0; i < size; i += len) {
+          for (int j = 0; j < half_len; ++j) {
+            int tw_idx = (dir == NTTDir::kForward) ? j * step : domain_max_size - j * step;
+            E u = current_elements[i + j];
+            E v = current_elements[i + j + half_len];
+            current_elements[i + j] = u + v;
+            current_elements[i + j + half_len] = (u - v) * twiddles[tw_idx];
+          }
+        }
+      }
+    }
+  }
+
+  template <typename E = scalar_t>
+  void transpose(const E* input, E* output, int rows, int cols)
+  {
+    for (int col = 0; col < cols; ++col) {
+      for (int row = 0; row < rows; ++row) {
+        output[col * rows + row] = input[row * cols + col];
+      }
+    }
+  }
+
+  template <typename S = scalar_t, typename E = scalar_t>
+  eIcicleError coset_mul(
+    int logn,
+    int domain_max_size,
+    E* elements,
+    int batch_size,
+    const S* twiddles = nullptr,
+    int stride = 0,
+    const S* arbitrary_coset = nullptr,
+    bool bit_rev = false,
+    NTTDir dir = NTTDir::kForward,
+    bool columns_batch = false)
+  {
+    int size = 1 << logn;
+    int idx;
+    for (int batch = 0; batch < batch_size; ++batch) {
+      E* current_elements = elements + batch * size;
+      if (arbitrary_coset != nullptr) {
+        for (int i = 1; i < size; ++i) {
+          idx = columns_batch ? batch : i;
+          idx = bit_rev ? bit_reverse(idx, logn) : idx;
+          current_elements[i] = current_elements[i] * arbitrary_coset[idx];
+        }
+      } else if (stride != 0) {
+        for (int i = 1; i < size; ++i) {
+          idx = bit_rev ? stride * (bit_reverse(i, logn)) : stride * i;
+          idx = dir == NTTDir::kForward ? idx : domain_max_size - idx;
+          current_elements[i] = current_elements[i] * twiddles[idx];
+        }
+      }
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename S = scalar_t, typename E = scalar_t>
+  eIcicleError cpu_ntt_ref(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output)
+  {
+    if (size & (size - 1)) { return eIcicleError::INVALID_ARGUMENT; }
```

Comment:
use ICICLE_LOG_ERROR to log errors to user in case of invalid inputs

Reply:
fixed